pkg/virt-controller/watch/application.go

```diff
@@ -89,14 +104,58 @@ func (vca *VirtControllerApp) Run() {
 	stop := make(chan struct{})
 	defer close(stop)
 	vca.informerFactory.Start(stop)
-	go vca.vmController.Run(3, stop)
-	//FIXME when we have more than one worker, we need a lock on the VM
-	go vca.migrationController.Run(3, stop)
-	httpLogger := logger.With("service", "http")
-	httpLogger.Info().Log("action", "listening", "interface", vca.host, "port", vca.port)
-	if err := http.ListenAndServe(vca.host+":"+strconv.Itoa(vca.port), nil); err != nil {
+
+	run := func(stopCh <-chan struct{}) {
+		go vca.vmController.Run(3, stop)
+		//FIXME when we have more than one worker, we need a lock on the VM
+		go vca.migrationController.Run(3, stop)
+		httpLogger := logger.With("service", "http")
+		httpLogger.Info().Log("action", "listening", "interface", vca.host, "port", vca.port)
+		if err := http.ListenAndServe(vca.host+":"+strconv.Itoa(vca.port), nil); err != nil {
```

Comment:
I believe we want the logger to run regardless of which instance is currently the leader.

Reply:
+1 we need the server to always run. I would say that we need two REST based checks. One should be /healthz (alrady exists, if I remember that correctly), and one we e.g. call /ready.

If we get elected, ready should return 200. Otherwise e.g. 503. This way we can make sure, that the k8s service always redirects the traffic to the one ready pod, if we add a rediness check, based on this endpoint.

The health endpoint would be mapped to a health check on the pod, so that k8s knows if something is obviously wrong, and restarts the pod.

Reply:
@davidvossel @rmohr For me it looks a little bit problematic because of both
`http.ListenAndServe` and `leaderElector.Run()` run in the infinitive loop until failure, so if I run some of some of them, the second does not run at all. And I can not check leader election status before I run `leaderElector.Run()`.
k8s use kube-api service for healthz checking, so they do not have such sort of problem.
Any thought?

Reply:
@cynepco3hahue You can e.g. run the http server in another go routine. If you get elected you can e.g. switch a boolean to yes from no and reflect that in the ready state. Does this make sense?

Reply:
I can see that in my setup the kubernetes-controller-manager has a livenessProbe configured.